<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>lassosum - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "lassosum";
    var mkdocs_page_input_path = "lassosum.md";
    var mkdocs_page_url = "/lassosum/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../base/">1. QC of Summary Statistics</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. Target Genotype QC</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Running PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">lassosum</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#running-prs-analysis">Running PRS analysis</a></li>
    

    </ul>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>3. Running PRS &raquo;</li>
        
      
    
    <li>lassosum</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/lassosum.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p><code>lassosum</code> is an <code>R</code> package for PRS calculation. 
It uses LASSO/Elastic Net estimates rather than p-value thresholding to generate PRS and are 
expected to provide a higher <span class="arithmatex">\(R^2\)</span> when compared to p-value thresholding.</p>
<p>You can install <code>lassosum</code> and its dependencies in <code>R</code> with the following</p>
<div class="codehilite"><pre><span></span>install.packages<span class="p">(</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;devtools&quot;</span><span class="p">,</span><span class="s">&quot;RcppArmadillo&quot;</span><span class="p">,</span> <span class="s">&quot;data.table&quot;</span><span class="p">,</span> <span class="s">&quot;Matrix&quot;</span><span class="p">),</span> dependencies<span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
install_github<span class="p">(</span><span class="s">&quot;tshmak/lassosum&quot;</span><span class="p">)</span>
</pre></div>

<p>Assuming we have the following files</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.valid.sample</strong></td>
<td align="center">This file contains the samples that passed all the QC</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.covariate</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the PCs of the samples</td>
</tr>
</tbody>
</table>
<h1 id="running-prs-analysis">Running PRS analysis<a class="headerlink" href="#running-prs-analysis" title="Permanent link">&para;</a></h1>
<p>We can run lassosum as follow</p>
<div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>lassosum<span class="p">)</span>
<span class="c1"># Prefer to work with data.table as it speeds up file reading</span>
<span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
<span class="kn">library</span><span class="p">(</span>methods<span class="p">)</span>
<span class="c1"># We like to use dplyr for it makes codes much more readable</span>
<span class="kn">library</span><span class="p">(</span>dplyr<span class="p">)</span>
sum.stat <span class="o">&lt;-</span> <span class="s">&quot;Height.QC.gz&quot;</span>
ref.bfile <span class="o">&lt;-</span> <span class="s">&quot;EUR.QC&quot;</span>
bfile <span class="o">&lt;-</span> <span class="s">&quot;EUR.QC&quot;</span>
<span class="c1"># Read in and process the covariates</span>
covariate <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.covariate&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
pcs <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.eigenvec&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="kp">colnames</span><span class="p">(</span>pcs<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">))</span>
cov <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>covariate<span class="p">,</span> pcs<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">))</span>

<span class="c1"># We will need the EUR.hg19 file provided by lassosum </span>
<span class="c1"># which are LD regions defined in Berisa and Pickrell (2015) for the European population and the hg19 genome.</span>
ld.file <span class="o">&lt;-</span>  <span class="s">&quot;EUR.hg19&quot;</span> 
<span class="c1"># Number of sample in base</span>
size <span class="o">&lt;-</span> <span class="m">253288</span>
<span class="c1"># output prefix</span>
prefix <span class="o">&lt;-</span> <span class="s">&quot;EUR&quot;</span>
<span class="c1"># Read in the target phenotype file</span>
fread<span class="p">(</span><span class="s">&quot;EUR.height&quot;</span><span class="p">,</span> data.table <span class="o">=</span> <span class="bp">F</span><span class="p">)</span> <span class="o">%&gt;%</span>
    select<span class="p">(</span>FID<span class="p">,</span> IID<span class="p">,</span> Pheno<span class="p">)</span> <span class="o">-&gt;</span> target.pheno
<span class="c1"># Read in samples to include in the analysis</span>
fread<span class="p">(</span><span class="s">&quot;EUR.valid.sample&quot;</span><span class="p">,</span> data.table<span class="o">=</span><span class="bp">F</span><span class="p">)</span><span class="o">%&gt;%</span>
    select<span class="p">(</span>FID<span class="p">,</span> IID<span class="p">)</span> <span class="o">-&gt;</span> target.keep
<span class="c1"># Read in the summary statistics</span>
ss <span class="o">&lt;-</span> fread<span class="p">(</span>sum.stat<span class="p">)</span>
<span class="c1"># Remove P-value = 0, which causes problem in the transformation</span>
ss <span class="o">&lt;-</span> ss<span class="p">[</span><span class="o">!</span>P <span class="o">==</span> <span class="m">0</span><span class="p">,</span> <span class="p">]</span>
<span class="c1"># Read in the LD blocks</span>
ld <span class="o">&lt;-</span> fread<span class="p">(</span>ld.file<span class="p">)</span>
<span class="c1"># Transform the P-values into correlation</span>
cor <span class="o">&lt;-</span> p2cor<span class="p">(</span>p <span class="o">=</span> ss<span class="o">$</span>P<span class="p">,</span>
        n <span class="o">=</span> size<span class="p">,</span>
        sign <span class="o">=</span> <span class="kp">log</span><span class="p">(</span>ss<span class="o">$</span>OR<span class="p">)</span>
        <span class="p">)</span>
<span class="c1"># Run the lassosum pipeline</span>
out <span class="o">&lt;-</span> lassosum.pipeline<span class="p">(</span>
    cor <span class="o">=</span> cor<span class="p">,</span>
    chr <span class="o">=</span> ss<span class="o">$</span>CHR<span class="p">,</span>
    pos <span class="o">=</span> ss<span class="o">$</span>BP<span class="p">,</span>
    A1 <span class="o">=</span> ss<span class="o">$</span>A1<span class="p">,</span>
    A2 <span class="o">=</span> ss<span class="o">$</span>A2<span class="p">,</span>
    ref.bfile <span class="o">=</span> ref.bfile<span class="p">,</span>
    keep.ref <span class="o">=</span> target.keep<span class="p">,</span>
    test.bfile <span class="o">=</span> bfile<span class="p">,</span>
    keep.test <span class="o">=</span> target.keep<span class="p">,</span>
    LDblocks <span class="o">=</span> ld<span class="p">,</span>
    trace <span class="o">=</span> <span class="m">2</span>
<span class="p">)</span>
<span class="c1"># Store the R2 results</span>
target.res <span class="o">&lt;-</span> validate<span class="p">(</span>out<span class="p">,</span> pheno <span class="o">=</span> target.pheno<span class="p">,</span> covar<span class="o">=</span>cov<span class="p">)</span>
<span class="c1"># Get the maximum R2</span>
r2 <span class="o">&lt;-</span> <span class="kp">max</span><span class="p">(</span>target.res<span class="o">$</span>validation.table<span class="o">$</span>value<span class="p">)</span><span class="o">^</span><span class="m">2</span>
</pre></div>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../ldpred/" class="btn btn-neutral" title="LDpred"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../ldpred/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
