<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Background - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Background";
    var mkdocs_page_input_path = "plink.md";
    var mkdocs_page_url = "/plink/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../base/">Base</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../lassosum/">Lassosum</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../ldpred/">Ldpred</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Background</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#background">Background</a></li>
    

    <li class="toctree-l2"><a href="#required-data">Required Data</a></li>
    

    <li class="toctree-l2"><a href="#remove-ambiguous-snps">Remove Ambiguous SNPs</a></li>
    

    <li class="toctree-l2"><a href="#strand-flipping">Strand Flipping</a></li>
    

    <li class="toctree-l2"><a href="#change-all-alleles-to-upper-case-for-easy-comparison">Change all alleles to upper case for easy comparison</a></li>
    

    <li class="toctree-l2"><a href="#function-for-finding-the-complementary-allele">Function for finding the complementary allele</a></li>
    

    <li class="toctree-l2"><a href="#get-snps-that-has-the-same-alleles-across-base-and-target">Get SNPs that has the same alleles across base and target</a></li>
    

    <li class="toctree-l2"><a href="#identify-snps-that-are-complementary-between-base-and-target">Identify SNPs that are complementary between base and target</a></li>
    

    <li class="toctree-l2"><a href="#update-these-allele-coding-in-the-bim-file">Update these allele coding in the bim file</a></li>
    

    <li class="toctree-l2"><a href="#identify-snps-that-need-flipping">identify SNPs that need flipping</a></li>
    

    <li class="toctree-l2"><a href="#identify-snps-that-need-flipping-complement">identify SNPs that need flipping &amp; complement</a></li>
    

    <li class="toctree-l2"><a href="#update-these-allele-coding-in-the-bim-file_1">Update these allele coding in the bim file</a></li>
    

    <li class="toctree-l2"><a href="#get-list-of-snps-that-need-to-change-the-a1-encoding">Get list of SNPs that need to change the A1 encoding</a></li>
    

    <li class="toctree-l2"><a href="#and-we-want-to-remove-any-snps-that-do-not-match-with-the-base-data">And we want to remove any SNPs that do not match with the base data</a></li>
    

    <li class="toctree-l2"><a href="#update-effect-size">Update Effect Size</a></li>
    

    <li class="toctree-l2"><a href="#clumping">Clumping</a></li>
    

    <li class="toctree-l2"><a href="#generate-prs">Generate PRS</a></li>
    

    <li class="toctree-l2"><a href="#accounting-for-population-stratification">Accounting for Population Stratification</a></li>
    

    <li class="toctree-l2"><a href="#finding-the-best-p-value-threshold">Finding the "Best" P-value threshold</a></li>
    

    <li class="toctree-l2"><a href="#read-in-the-phenotype-file">Read in the phenotype file</a></li>
    

    <li class="toctree-l2"><a href="#read-in-the-pcs">Read in the PCs</a></li>
    

    <li class="toctree-l2"><a href="#the-default-output-from-plink-does-not-include-a-header">The default output from plink does not include a header</a></li>
    

    <li class="toctree-l2"><a href="#to-make-things-simple-we-will-add-the-appropriate-headers">To make things simple, we will add the appropriate headers</a></li>
    

    <li class="toctree-l2"><a href="#16-because-there-are-6-pcs">(1:6 because there are 6 PCs)</a></li>
    

    <li class="toctree-l2"><a href="#read-in-the-covariates-here-it-is-sex">Read in the covariates (here, it is sex)</a></li>
    

    <li class="toctree-l2"><a href="#now-merge-the-files">Now merge the files</a></li>
    

    <li class="toctree-l2"><a href="#we-can-then-calculate-the-null-model-model-with-prs-using-a-linear-regression">We can then calculate the null model (model with PRS) using a linear regression</a></li>
    

    <li class="toctree-l2"><a href="#as-height-is-quantitative">(as height is quantitative)</a></li>
    

    <li class="toctree-l2"><a href="#and-the-r2-of-the-null-model-is">And the R2 of the null model is</a></li>
    

    <li class="toctree-l2"><a href="#best-result-is">Best result is:</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">Plink visual</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../prsice/">Prsice</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">Target</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Background</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/plink.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p>In this section, we will try to calculate polygenic risk score using <code>plink</code> to illustrate some common procedure
performed by PRS software.</p>
<h1 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h1>
<p>In previous sections, we have generated the following files</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.valid.sample</strong></td>
<td align="center">This file contains the samples that passed all the QC</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.covariate</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
</tbody>
</table>
<h1 id="remove-ambiguous-snps">Remove Ambiguous SNPs<a class="headerlink" href="#remove-ambiguous-snps" title="Permanent link">&para;</a></h1>
<p>If the base and target data were generated using different genotyping chips and the chromosome strand (+/-) for either is unknown, then it is not possible to match ambiguous SNPs (i.e. those with complementary alleles, either C/G or A/T) across the data sets, because it will be unknown whether the base and target data are referring to the same allele or not. </p>
<p>Ambiguous SNPs can be obtained by examining the bim file:
<div class="codehilite"><pre><span></span>awk <span class="s1">&#39;!( ($5==&quot;A&quot; &amp;&amp; $6==&quot;T&quot;) || \</span>
<span class="s1">        ($5==&quot;T&quot; &amp;&amp; $6==&quot;A&quot;) || \</span>
<span class="s1">        ($5==&quot;G&quot; &amp;&amp; $6==&quot;C&quot;) || \</span>
<span class="s1">        ($5==&quot;C&quot; &amp;&amp; $6==&quot;G&quot;)) {print}&#39;</span> <span class="se">\</span>
        EUR.QC.bim &gt; EUR.unambig.snp 
</pre></div></p>
<details class="note"><summary>How many ambiguous SNPs were there?</summary><p>There are <code>330,818</code> ambiguous SNPs</p></details><h1 id="strand-flipping">Strand Flipping<a class="headerlink" href="#strand-flipping" title="Permanent link">&para;</a></h1>
<p>In addition, when there are non-ambiguous mismatch in allele 
coding between the data sets, such as A/C in the base
and G/T in the target data, then this can be resolved by 
‘flipping’ the alleles in the target data to their complementary alleles. 
This has to be done with mulitpe steps</p>
<ol>
<li>Get the correct A1 alleles for the bim file</li>
</ol>
<p>```R tab="Without data.table"
bim &lt;- read.table("EUR.QC.bim", header = F, stringsAsFactors = F)
colnames(bim) &lt;- c("CHR", "SNP", "CM", "BP", "B.A1", "B.A2")
height &lt;-
    read.table(gzfile("Height.QC.gz"),
               header = T,
               stringsAsFactors = F)</p>
<h1 id="change-all-alleles-to-upper-case-for-easy-comparison">Change all alleles to upper case for easy comparison<a class="headerlink" href="#change-all-alleles-to-upper-case-for-easy-comparison" title="Permanent link">&para;</a></h1>
<p>height<span class="arithmatex">\(A1 &lt;- toupper(height\)</span>A1)
height<span class="arithmatex">\(A2 &lt;- toupper(height\)</span>A2)
bim<span class="arithmatex">\(B.A1 &lt;- toupper(bim\)</span>B.A1)
bim<span class="arithmatex">\(B.A2 &lt;- toupper(bim\)</span>B.A2)
info &lt;- merge(bim, height, by = c("SNP", "CHR", "BP"))</p>
<h1 id="function-for-finding-the-complementary-allele">Function for finding the complementary allele<a class="headerlink" href="#function-for-finding-the-complementary-allele" title="Permanent link">&para;</a></h1>
<p>complement &lt;- function(x) {
    switch (
        x,
        "A" = "T",
        "C" = "G",
        "T" = "A",
        "G" = "C",
        return(NA)
    )
}</p>
<h1 id="get-snps-that-has-the-same-alleles-across-base-and-target">Get SNPs that has the same alleles across base and target<a class="headerlink" href="#get-snps-that-has-the-same-alleles-across-base-and-target" title="Permanent link">&para;</a></h1>
<p>info.match &lt;- subset(info, A1 == B.A1 &amp; A2 == B.A2)</p>
<h1 id="identify-snps-that-are-complementary-between-base-and-target">Identify SNPs that are complementary between base and target<a class="headerlink" href="#identify-snps-that-are-complementary-between-base-and-target" title="Permanent link">&para;</a></h1>
<p>info<span class="arithmatex">\(C.A1 &lt;- sapply(info\)</span>B.A1, complement)
info<span class="arithmatex">\(C.A2 &lt;- sapply(info\)</span>B.A2, complement)
info.complement &lt;- subset(info, A1 == C.A1 &amp; A2 == C.A2)</p>
<h1 id="update-these-allele-coding-in-the-bim-file">Update these allele coding in the bim file<a class="headerlink" href="#update-these-allele-coding-in-the-bim-file" title="Permanent link">&para;</a></h1>
<p>bim[bim<span class="arithmatex">\(SNP %in% info.complement\)</span>SNP,]<span class="arithmatex">\(B.A1 &lt;-
    sapply(bim[bim\)</span>SNP %in% info.complement<span class="arithmatex">\(SNP,]\)</span>B.A1, complement)
bim[bim<span class="arithmatex">\(SNP %in% info.complement\)</span>SNP,]<span class="arithmatex">\(B.A2 &lt;-
    sapply(bim[bim\)</span>SNP %in% info.complement<span class="arithmatex">\(SNP,]\)</span>B.A2, complement)</p>
<h1 id="identify-snps-that-need-flipping">identify SNPs that need flipping<a class="headerlink" href="#identify-snps-that-need-flipping" title="Permanent link">&para;</a></h1>
<p>info.flip &lt;- subset(info, A1 == B.A2 &amp; A2 == B.A1)</p>
<h1 id="identify-snps-that-need-flipping-complement">identify SNPs that need flipping &amp; complement<a class="headerlink" href="#identify-snps-that-need-flipping-complement" title="Permanent link">&para;</a></h1>
<p>info.cflip &lt;- subset(info, A1 == C.A2 &amp; A2 == C.A1)</p>
<h1 id="update-these-allele-coding-in-the-bim-file_1">Update these allele coding in the bim file<a class="headerlink" href="#update-these-allele-coding-in-the-bim-file_1" title="Permanent link">&para;</a></h1>
<p>com.snps &lt;- bim<span class="arithmatex">\(SNP %in% info.cflip\)</span>SNP
bim[com.snps,]<span class="arithmatex">\(B.A1 &lt;- sapply(bim[com.snps,]\)</span>B.A1, complement)
bim[com.snps,]<span class="arithmatex">\(B.A2 &lt;- sapply(bim[com.snps,]\)</span>B.A2, complement)</p>
<h1 id="get-list-of-snps-that-need-to-change-the-a1-encoding">Get list of SNPs that need to change the A1 encoding<a class="headerlink" href="#get-list-of-snps-that-need-to-change-the-a1-encoding" title="Permanent link">&para;</a></h1>
<p>flip &lt;- rbind(info.flip, info.cflip)
flip.snp &lt;- data.frame(SNP = flip<span class="arithmatex">\(SNP, A1 = flip\)</span>A1)
write.table(flip.snp,
            "EUR.update.a1",
            quote = F,
            row.names = F)
write.table(
    bim,
    "EUR.QC.adj.bim",
    quote = F,
    row.names = F,
    col.names = F
)</p>
<h1 id="and-we-want-to-remove-any-snps-that-do-not-match-with-the-base-data">And we want to remove any SNPs that do not match with the base data<a class="headerlink" href="#and-we-want-to-remove-any-snps-that-do-not-match-with-the-base-data" title="Permanent link">&para;</a></h1>
<p>mismatch &lt;-
    bim<span class="arithmatex">\(SNP[!(bim\)</span>SNP %in% info.match<span class="arithmatex">\(SNP |
                  bim\)</span>SNP %in% info.complement<span class="arithmatex">\(SNP | 
                  bim\)</span>SNP %in% flip$SNP)]
write.table(
    mismatch,
    "EUR.mismatch",
    quote = F,
    row.names = F,
    col.names = F
)
<div class="codehilite"><pre><span></span><span class="sb">``</span>`R tab<span class="o">=</span><span class="s">&quot;With data.table&quot;</span>
<span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
bim <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.QC.bim&quot;</span><span class="p">)</span>
bim.col <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CM&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span> <span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span>
setnames<span class="p">(</span>bim<span class="p">,</span> <span class="kp">colnames</span><span class="p">(</span>bim<span class="p">),</span> bim.col<span class="p">)</span>
height <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">)</span>
<span class="c1"># Change all alleles to upper case for easy comparison</span>
height<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;A1&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="p">)</span><span class="o">:=</span><span class="kt">list</span><span class="p">(</span><span class="kp">toupper</span><span class="p">(</span>A1<span class="p">),</span> <span class="kp">toupper</span><span class="p">(</span>A2<span class="p">))]</span>
bim<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span><span class="s">&quot;B.A2&quot;</span><span class="p">)</span><span class="o">:=</span><span class="kt">list</span><span class="p">(</span><span class="kp">toupper</span><span class="p">(</span>B.A1<span class="p">),</span> <span class="kp">toupper</span><span class="p">(</span>B.A2<span class="p">))]</span>
info <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>bim<span class="p">,</span> height<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">))</span>
<span class="c1"># Function for calculating the complementary allele</span>
complement <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">){</span>
    <span class="kr">switch</span> <span class="p">(</span>x<span class="p">,</span>
        <span class="s">&quot;A&quot;</span> <span class="o">=</span> <span class="s">&quot;T&quot;</span><span class="p">,</span>
        <span class="s">&quot;C&quot;</span> <span class="o">=</span> <span class="s">&quot;G&quot;</span><span class="p">,</span>
        <span class="s">&quot;T&quot;</span> <span class="o">=</span> <span class="s">&quot;A&quot;</span><span class="p">,</span>
        <span class="s">&quot;G&quot;</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">,</span>
        <span class="kr">return</span><span class="p">(</span><span class="kc">NA</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">}</span>
<span class="c1"># Identify SNPs that are complementary between base and target</span>
com.snps <span class="o">&lt;-</span> info<span class="p">[</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A1 <span class="o">&amp;</span>
                     <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A2<span class="p">,</span> SNP<span class="p">]</span>
<span class="c1"># Now update the bim file</span>
bim<span class="p">[</span>SNP <span class="o">%in%</span> com.snps<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># identify SNPs that need flipping &amp; complement</span>
com.flip <span class="o">&lt;-</span> info<span class="p">[</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A2 <span class="o">&amp;</span>
                     <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A1<span class="p">,</span> SNP<span class="p">]</span>
<span class="c1"># Now update the bim file</span>
bim<span class="p">[</span>SNP <span class="o">%in%</span> com.flip<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># Obtain list of SNPs that require flipping</span>
flip <span class="o">&lt;-</span> info<span class="p">[</span>B.A1<span class="o">==</span>A2 <span class="o">&amp;</span> B.A2<span class="o">==</span>A1<span class="p">]</span>
<span class="c1"># Now generate file for PLINK </span>
fwrite<span class="p">(</span>flip<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;A1&quot;</span><span class="p">)],</span> <span class="s">&quot;EUR.update.a1&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="c1"># Write the updated bim file</span>
fwrite<span class="p">(</span>bim<span class="p">,</span> <span class="s">&quot;EUR.QC.adj.bim&quot;</span><span class="p">,</span> col.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="c1"># We can then remove all mismatch SNPs</span>
matched <span class="o">&lt;-</span> info<span class="p">[(</span>A1 <span class="o">==</span> B.A1 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A2<span class="p">)</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> B.A2 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A1<span class="p">)</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">&amp;</span>
                         A2 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">&amp;</span>
                         A2 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">))]</span>
mismatch <span class="o">&lt;-</span> bim<span class="p">[</span><span class="o">!</span>SNP<span class="o">%in%</span>matched<span class="o">$</span>SNP<span class="p">,</span> SNP<span class="p">]</span>
write.table<span class="p">(</span>mismatch<span class="p">,</span> <span class="s">&quot;EUR.mismatch&quot;</span><span class="p">,</span> quote<span class="o">=</span><span class="bp">F</span><span class="p">,</span> row.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> col.names<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</pre></div></p>
<p>The above script will generate three files: <strong>EUR.QC.adj.bim</strong>, <strong>EUR.update.a1</strong> and <strong>EUR.mismatch</strong>. We want to replace
<strong>EUR.QC.bim</strong> with <strong>EUR.QC.adj.bim</strong>:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Make a back up</span>
mv EUR.QC.bim EUR.QC.bim.bk
ln -s EUR.QC.adj.bim EUR.QC.bim
</pre></div>

<p>We can then generate a new genotype file with the correct genetic encodings
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC <span class="se">\</span>
    --a1-allele EUR.update.a1 <span class="se">\</span>
    --make-bed <span class="se">\</span>
    --keep EUR.valid.sample <span class="se">\</span>
    --extract EUR.unambig.snp <span class="se">\</span>
    --exclude EUR.mismatch <span class="se">\</span>
    --out EUR.QC.flipped
</pre></div></p>
<h1 id="update-effect-size">Update Effect Size<a class="headerlink" href="#update-effect-size" title="Permanent link">&para;</a></h1>
<p>When odd ratios (OR) instead of BETA are provided, the PRS might have to calculated using a multiplicative model.
To simplify the calculation, we usually take the natural logarithm of the OR such that an additive model can be applied. 
We can obtain the transformed summary statistics with <code>R</code>:</p>
<p>```R tab="Without data.table"
dat &lt;- read.table(gzfile("Height.QC.gz"), header=T)
dat<span class="arithmatex">\(OR &lt;- log(dat\)</span>OR)
write.table(dat, "Height.QC.Transformed", quote=F, row.names=F)
<div class="codehilite"><pre><span></span><span class="sb">``</span>`R tab<span class="o">=</span><span class="s">&quot;With data.table&quot;</span>
<span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
dat <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">)</span>
fwrite<span class="p">(</span>dat<span class="p">[,</span>OR<span class="o">:=</span><span class="kp">log</span><span class="p">(</span>OR<span class="p">)],</span> <span class="s">&quot;Height.QC.Transformed&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It might be tempting to perform the log transofrmation using <code>awk</code>.
However, due to a lower arithmetic precision of <code>awk</code>, less accurate results
might be obtained. 
Therefore it is best to do the transformation in <code>R</code> or allow the PRS software to perform the transformation directly. </p>
</div>
<h1 id="clumping">Clumping<a class="headerlink" href="#clumping" title="Permanent link">&para;</a></h1>
<p>Linkage disequilibrium introduce a strong correlation structure across the genome, makes identifying the independent
genetic effects extremely challenging. 
One simple method is to perform clumping, which preferentially selects SNPs most
associated with the trait under study when removing SNPs in LD. </p>
<p><div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --clump-p1 <span class="m">1</span> <span class="se">\</span>
    --clump-r2 <span class="m">0</span>.1 <span class="se">\</span>
    --clump-kb <span class="m">250</span> <span class="se">\</span>
    --clump Height.QC.transformed <span class="se">\</span>
    --clump-snp-field SNP <span class="se">\</span>
    --clump-field P <span class="se">\</span>
    --out EUR
</pre></div>
Each of the new parameters corresponds to the following</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">clump-p1</td>
<td align="center">1</td>
<td align="left">P-value threshold for a SNP to be included as an index SNP. 1 is selected such that all SNPs are include for clumping</td>
</tr>
<tr>
<td align="center">clump-r2</td>
<td align="center">0.1</td>
<td align="left">SNPs having <span class="arithmatex">\(r^2\)</span> higher than 0.1 with the index SNPs will be removed</td>
</tr>
<tr>
<td align="center">clump-kb</td>
<td align="center">250</td>
<td align="left">SNPs within 250k of the index SNP are considered for clumping</td>
</tr>
<tr>
<td align="center">clump</td>
<td align="center">Height.QC.transformed</td>
<td align="left">Summary statistic file containing the p-value information</td>
</tr>
<tr>
<td align="center">clump-snp-field</td>
<td align="center">SNP</td>
<td align="left">Specify that the column <code>SNP</code> contains the SNP IDs</td>
</tr>
<tr>
<td align="center">clump-field</td>
<td align="center">P</td>
<td align="left">Specify that the column <code>P</code> contains the P-value information</td>
</tr>
</tbody>
</table>
<p>A more detailed document can be found <a href="https://www.cog-genomics.org/plink/1.9/postproc#clump">here</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="arithmatex">\(r^2\)</span> values computed by <code>--clump</code> are based on maximum likelihood haplotype frequency estimates</p>
</div>
<p>This will generate <strong>EUR.clumped</strong>, containing the index SNPs after clumping is performed.
We can extract the index SNP ID by doing</p>
<div class="codehilite"><pre><span></span>awk <span class="s1">&#39;NR!=1{print $3}&#39;</span> EUR.clumped &gt;  EUR.valid.snp
</pre></div>

<blockquote>
<p><code>$3</code> because the third column contains the SNP ID</p>
</blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your target sample is small (e.g. &lt;500), you can try using the 1000 Genome samples for the LD calculation.
Make sure you use the population that best represents your sample.</p>
</div>
<h1 id="generate-prs">Generate PRS<a class="headerlink" href="#generate-prs" title="Permanent link">&para;</a></h1>
<p><code>plink</code> provide a handy function <code>--score</code> and <code>--q-score-range</code> for calculating polygenic score.</p>
<p>We will need three files</p>
<ol>
<li>The summary statistic file: <strong>Height.QC.Transformed</strong></li>
<li>A file containing SNP ID and their corresponding p-value (<code>$1</code> because SNP ID is located at the first column; <code>$8</code> because P-value is located at the eigth column)
<div class="codehilite"><pre><span></span>awk <span class="s1">&#39;{print $1,$8}&#39;</span> Height.QC.Transformed &gt; SNP.pvalue
</pre></div></li>
<li>A file containing the P-value thresholds we are testing. Here we will only test a few for illustration purposes
<div class="codehilite"><pre><span></span><span class="nb">echo</span> <span class="s2">&quot;0.001 0 0.001&quot;</span> &gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.05 0 0.05&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.1 0 0.1&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.2 0 0.2&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.3 0 0.3&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.4 0 0.4&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.5 0 0.5&quot;</span> &gt;&gt; range_list
</pre></div>
The format of the <strong>range_list</strong> file should be as follow</li>
</ol>
<table>
<thead>
<tr>
<th align="center">Name of Threshold</th>
<th align="center">Lower bound</th>
<th align="center">Upper Bound</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The boundary are inclusive. For example, for the <code>0.05</code> threshold, we include all SNPs with P-value from 
<code>0</code> to <code>0.05</code>, <strong>including</strong> any SNPs with P-value equal to <code>0.05</code></p>
</div>
<p>We can then calculate the PRS with the following <code>plink</code> command:</p>
<p><div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --extract EUR.valid.snp <span class="se">\</span>
    --score Height.QC.Transformed <span class="m">1</span> <span class="m">4</span> <span class="m">11</span> header <span class="se">\</span>
    --q-score-range range_list SNP.pvalue <span class="se">\</span>
    --out EUR
</pre></div>
Meaning of the new parameters are as follow</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">score</td>
<td align="center">Height.QC.Transformed 1 4 11 header</td>
<td align="left">We read from the <strong>Height.QC.Transformed</strong> file, assuming the <code>1</code>st column to be the SNP ID; <code>4</code>th column to be the effective allele information; <code>11</code>th column to be the effect size estimate; and the file contains a <code>header</code></td>
</tr>
<tr>
<td align="center">q-score-range</td>
<td align="center">range_test SNP.pvalue</td>
<td align="left">We want to calculate PRS based on the thresholds defined in <strong>range_test</strong>, where the threshold values (p-values) were stored in <strong>SNP.pvalue</strong></td>
</tr>
</tbody>
</table>
<p>The above command and range_list will generate 7 files:</p>
<ol>
<li>EUR.0.5.profile</li>
<li>EUR.0.4.profile</li>
<li>EUR.0.3.profile</li>
<li>EUR.0.2.profile</li>
<li>EUR.0.1.profile</li>
<li>EUR.0.05.profile</li>
<li>EUR.0.001.profile</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The default formular for PRS calculation in PLINK is:
(Assuming the effect size of SNP <span class="arithmatex">\(i\)</span> is <span class="arithmatex">\(S_i\)</span>;  the number of effective allele observed in sample <span class="arithmatex">\(j\)</span> is <span class="arithmatex">\(G_{ij}\)</span>; the ploidy of the sample is <span class="arithmatex">\(P\)</span> (It should be 2 for human); the number of samples included in the PRS be <span class="arithmatex">\(N\)</span>; and the number of non-missing SNPs observed in sample <span class="arithmatex">\(j\)</span> be <span class="arithmatex">\(M_j\)</span>)
$$
PRS_j =\frac{ \sum_i^NS_i*G_{ij}}{P*M_j}
$$</p>
<p>If sample has a missing genotype for SNP <span class="arithmatex">\(i\)</span>, the population minor allele frequency times ploidy (<span class="arithmatex">\(MAF_i*P\)</span>) is used inplace of <span class="arithmatex">\(G_{ij}\)</span></p>
</div>
<h1 id="accounting-for-population-stratification">Accounting for Population Stratification<a class="headerlink" href="#accounting-for-population-stratification" title="Permanent link">&para;</a></h1>
<p>Population structure is the principal source of confounding in GWAS and are usually accounted for by incorporating the principal components (PCs) as a covariate. 
Similarly, we can incorporate PCs in our PRS analysis to account for population stratification.</p>
<p>Again, we can calculate the PCs using <code>plink</code> 
<div class="codehilite"><pre><span></span><span class="c1"># First, we need to perform prunning</span>
plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --extract EUR.valid.snp <span class="se">\</span>
    --indep-pairwise <span class="m">200</span> <span class="m">50</span> <span class="m">0</span>.25 <span class="se">\</span>
    --out EUR
<span class="c1"># Then we calculate the first 6 PCs</span>
plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --extract EUR.prune.in <span class="se">\</span>
    --pca <span class="m">6</span> <span class="se">\</span>
    --out EUR
</pre></div></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>One way to select the appropriate number of PCs is to perform GWAS on the trait of interest with different number of PCs.
<a href="https://github.com/bulik/ldsc">LDSC</a> analysis can then be performed on each of the resulted GWAS summary statistics. 
By observing the estimated intercept, one can select the number of PCs that provide an intercept estimate closer to 1, which might
suggest a smaller influence of population stratification.</p>
</div>
<p>The eigen-vector (PCs) are stored in <strong>EUR.eigenvec</strong> and can be used as a covariate in the regression model to account for population stratification.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If the base and target samples are collected from different population (e.g. Caucasian vs African ), the results from PRS analysis will be biased (see <a href="https://www.ncbi.nlm.nih.gov/pubmed/28366442">Martin et al</a>).</p>
</div>
<h1 id="finding-the-best-p-value-threshold">Finding the "Best" P-value threshold<a class="headerlink" href="#finding-the-best-p-value-threshold" title="Permanent link">&para;</a></h1>
<p>The "best" p-value threshold for PRS construction are usually not known. 
To identify the "best" PRS, we can perform a regression between the calculated PRS and the 
sample phenotype and select the PRS that explains most of the phenotypic variation. 
This can be achieved using <code>R</code>.</p>
<p>```R tab="detail"
p.threshold &lt;- c(0.001,0.05,0.1,0.2,0.3,0.4,0.5)</p>
<h1 id="read-in-the-phenotype-file">Read in the phenotype file<a class="headerlink" href="#read-in-the-phenotype-file" title="Permanent link">&para;</a></h1>
<p>phenotype &lt;- read.table("EUR.height", header=T)</p>
<h1 id="read-in-the-pcs">Read in the PCs<a class="headerlink" href="#read-in-the-pcs" title="Permanent link">&para;</a></h1>
<p>pcs &lt;- read.table("EUR.eigenvec", header=F)</p>
<h1 id="the-default-output-from-plink-does-not-include-a-header">The default output from plink does not include a header<a class="headerlink" href="#the-default-output-from-plink-does-not-include-a-header" title="Permanent link">&para;</a></h1>
<h1 id="to-make-things-simple-we-will-add-the-appropriate-headers">To make things simple, we will add the appropriate headers<a class="headerlink" href="#to-make-things-simple-we-will-add-the-appropriate-headers" title="Permanent link">&para;</a></h1>
<h1 id="16-because-there-are-6-pcs">(1:6 because there are 6 PCs)<a class="headerlink" href="#16-because-there-are-6-pcs" title="Permanent link">&para;</a></h1>
<p>colnames(pcs) &lt;- c("FID", "IID", paste0("PC",1:6)) </p>
<h1 id="read-in-the-covariates-here-it-is-sex">Read in the covariates (here, it is sex)<a class="headerlink" href="#read-in-the-covariates-here-it-is-sex" title="Permanent link">&para;</a></h1>
<p>covariate &lt;- read.table("EUR.covariate", header=T)</p>
<h1 id="now-merge-the-files">Now merge the files<a class="headerlink" href="#now-merge-the-files" title="Permanent link">&para;</a></h1>
<p>pheno &lt;- merge(merge(phenotype, covariate, by=c("FID", "IID")), pcs, by=c("FID","IID"))</p>
<h1 id="we-can-then-calculate-the-null-model-model-with-prs-using-a-linear-regression">We can then calculate the null model (model with PRS) using a linear regression<a class="headerlink" href="#we-can-then-calculate-the-null-model-model-with-prs-using-a-linear-regression" title="Permanent link">&para;</a></h1>
<h1 id="as-height-is-quantitative">(as height is quantitative)<a class="headerlink" href="#as-height-is-quantitative" title="Permanent link">&para;</a></h1>
<p>null.model &lt;- lm(Height~., data=pheno[,!colnames(pheno)%in%c("FID","IID")])</p>
<h1 id="and-the-r2-of-the-null-model-is">And the R2 of the null model is<a class="headerlink" href="#and-the-r2-of-the-null-model-is" title="Permanent link">&para;</a></h1>
<p>null.r2 &lt;- summary(null.model)<span class="arithmatex">\(r.squared
prs.result &lt;- NULL
for(i in p.threshold){
    # Go through each p-value threshold
    prs &lt;- read.table(paste0("EUR.",i,".profile"), header=T)
    # Merge the prs with the phenotype matrix
    # We only want the FID, IID and PRS from the PRS file, therefore we only select the 
    # relevant columns
    pheno.prs &lt;- merge(pheno, prs[,c("FID","IID", "SCORE")], by=c("FID", "IID"))
    # Now perform a linear regression on Height with PRS and the covariates
    # ignoring the FID and IID from our model
    model &lt;- lm(Height~., data=pheno.prs[,!colnames(pheno.prs)%in%c("FID","IID")])
    # model R2 is obtained as 
    model.r2 &lt;- summary(model)\)</span>r.squared
    # R2 of PRS is simply calculated as the model R2 minus the null R2
    prs.r2 &lt;- model.r2-null.r2
    # We can also obtain the coeffcient and p-value of association of PRS as follow
    prs.coef &lt;- summary(model)$coeff["SCORE",]
    prs.beta &lt;- as.numeric(prs.coef[1])
    prs.se &lt;- as.numeric(prs.coef[2])
    prs.p &lt;- as.numeric(prs.coef[4])
    # We can then store the results
    prs.result &lt;- rbind(prs.result, data.frame(Threshold=i, R2=prs.r2, P=prs.p, BETA=prs.beta,SE=prs.se))
}</p>
<h1 id="best-result-is">Best result is:<a class="headerlink" href="#best-result-is" title="Permanent link">&para;</a></h1>
<p>prs.result[which.max(prs.result$R2),]
<div class="codehilite"><pre><span></span><span class="err">```</span><span class="n">R</span> <span class="n">tab</span><span class="o">=</span><span class="s">&quot;quick&quot;</span>
<span class="n">p</span><span class="p">.</span><span class="n">threshold</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.05</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.2</span><span class="p">,</span><span class="mf">0.3</span><span class="p">,</span><span class="mf">0.4</span><span class="p">,</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">phenotype</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="p">.</span><span class="n">table</span><span class="p">(</span><span class="s">&quot;EUR.height&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
<span class="n">pcs</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="p">.</span><span class="n">table</span><span class="p">(</span><span class="s">&quot;EUR.eigenvec&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
<span class="n">colnames</span><span class="p">(</span><span class="n">pcs</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span><span class="mi">1</span><span class="o">:</span><span class="mi">6</span><span class="p">))</span> 
<span class="n">covariate</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="p">.</span><span class="n">table</span><span class="p">(</span><span class="s">&quot;EUR.covariate&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
<span class="n">pheno</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">merge</span><span class="p">(</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">covariate</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">)),</span> <span class="n">pcs</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">))</span>
<span class="n">null</span><span class="p">.</span><span class="n">r2</span> <span class="o">&lt;-</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span><span class="n">Height</span><span class="o">~</span><span class="p">.,</span> <span class="n">data</span><span class="o">=</span><span class="n">pheno</span><span class="p">[,</span><span class="o">!</span><span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">)</span><span class="o">%</span><span class="k">in</span><span class="o">%</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)]))</span><span class="err">$</span><span class="n">r</span><span class="p">.</span><span class="n">squared</span>
<span class="n">prs</span><span class="p">.</span><span class="n">result</span> <span class="o">&lt;-</span> <span class="nb">NULL</span>
<span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="k">in</span> <span class="n">p</span><span class="p">.</span><span class="n">threshold</span><span class="p">){</span>
    <span class="n">pheno</span><span class="p">.</span><span class="n">prs</span> <span class="o">&lt;-</span> <span class="n">merge</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> 
                        <span class="n">read</span><span class="p">.</span><span class="n">table</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s">&quot;EUR.&quot;</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="s">&quot;.profile&quot;</span><span class="p">),</span> <span class="n">header</span><span class="o">=</span><span class="n">T</span><span class="p">)[,</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="s">&quot;SCORE&quot;</span><span class="p">)],</span>
                        <span class="n">by</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">))</span>
    <span class="n">model</span> <span class="o">&lt;-</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span><span class="n">Height</span><span class="o">~</span><span class="p">.,</span> <span class="n">data</span><span class="o">=</span><span class="n">pheno</span><span class="p">.</span><span class="n">prs</span><span class="p">[,</span><span class="o">!</span><span class="n">colnames</span><span class="p">(</span><span class="n">pheno</span><span class="p">.</span><span class="n">prs</span><span class="p">)</span><span class="o">%</span><span class="k">in</span><span class="o">%</span><span class="n">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)]))</span>
    <span class="n">model</span><span class="p">.</span><span class="n">r2</span> <span class="o">&lt;-</span> <span class="n">model</span><span class="err">$</span><span class="n">r</span><span class="p">.</span><span class="n">squared</span>
    <span class="n">prs</span><span class="p">.</span><span class="n">r2</span> <span class="o">&lt;-</span> <span class="n">model</span><span class="p">.</span><span class="n">r2</span><span class="o">-</span><span class="n">null</span><span class="p">.</span><span class="n">r2</span>
    <span class="n">prs</span><span class="p">.</span><span class="n">coef</span> <span class="o">&lt;-</span> <span class="n">model</span><span class="err">$</span><span class="n">coeff</span><span class="p">[</span><span class="s">&quot;SCORE&quot;</span><span class="p">,]</span>
    <span class="n">prs</span><span class="p">.</span><span class="n">result</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">result</span><span class="p">,</span> 
        <span class="n">data</span><span class="p">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Threshold</span><span class="o">=</span><span class="n">i</span><span class="p">,</span> <span class="n">R2</span><span class="o">=</span><span class="n">prs</span><span class="p">.</span><span class="n">r2</span><span class="p">,</span> 
                    <span class="n">P</span><span class="o">=</span><span class="n">as</span><span class="p">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">coef</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span> 
                    <span class="n">BETA</span><span class="o">=</span><span class="n">as</span><span class="p">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">coef</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                    <span class="n">SE</span><span class="o">=</span><span class="n">as</span><span class="p">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">coef</span><span class="p">[</span><span class="mi">2</span><span class="p">])))</span>
<span class="p">}</span>
<span class="n">print</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">result</span><span class="p">[</span><span class="n">which</span><span class="p">.</span><span class="n">max</span><span class="p">(</span><span class="n">prs</span><span class="p">.</span><span class="n">result</span><span class="err">$</span><span class="n">R2</span><span class="p">),])</span>
</pre></div></p>
<details class="note"><summary>Which p-value threshold generate the "best" PRS?</summary><p>0.05</p></details><details class="note"><summary>How much phenotypic variation does the "best" PRS explains?</summary><p>0.03921047</p></details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../plink_visual/" class="btn btn-neutral float-right" title="Plink visual">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../ldpred/" class="btn btn-neutral" title="Ldpred"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../ldpred/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../plink_visual/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
