<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>1. QC of Summary Statistics - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "1. QC of Summary Statistics";
    var mkdocs_page_input_path = "base.md";
    var mkdocs_page_url = "/base/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">1. QC of Summary Statistics</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#obtaining-the-summary-statistic-file">Obtaining the summary statistic file</a></li>
    

    <li class="toctree-l2"><a href="#reading-the-summary-statistic-file">Reading the summary statistic file</a></li>
    

    <li class="toctree-l2"><a href="#removing-duplicated-snps">Removing duplicated SNPs</a></li>
    

    <li class="toctree-l2"><a href="#filtering-snps-with-low-info-score-or-maf">Filtering SNPs with low INFO score or MAF</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. Target Genotype QC</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Running PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>1. QC of Summary Statistics</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/base.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>First step in Polygenic Risk Score (PRS) analyses is to obtain the GWAS summary statistics. </p>
<p>In this example, we will use a modified version of the Height GWAS summary statistics generated by the <a href="https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files#GWAS_Anthropometric_2014_Height">GIANT consortium</a></p>
<h1 id="obtaining-the-summary-statistic-file">Obtaining the summary statistic file<a class="headerlink" href="#obtaining-the-summary-statistic-file" title="Permanent link">&para;</a></h1>
<p>You can download the summary statistic file using the following script:
<div class="codehilite"><pre><span></span>curl https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz -L -O
</pre></div></p>
<p>which will create a file call <strong>GIANT.height.gz</strong> in your current directory. </p>
<p>A common problem is that the downloaded file can be corrupted, which can generate various error messages in the 
down-stream analyses. To avoid un-necessary waste of time, it is generally a good practice to check if the file is
intact. This can be done by performing the <code>md5sum</code> check:</p>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">Linux</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5sum GIANT.height.gz
</pre></div></div>
<input name="__tabs_1" type="radio" id="__tab_1_1" />
<label for="__tab_1_1">OS X</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5 GIANT.height.gz
</pre></div></div>
</div>
<p>if the file is intact, <code>md5sum</code> should generate a string of characters: <code>c79734b099cea663d2808bfde2e9a422</code>. 
If a different string is generated, the file is likely corrupted </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In most scenarios, a different <code>md5sum</code> is generated for different files. Therefore it is a nice simple way of 
checking whether the file is downloaded correctly. </p>
</div>
<h1 id="reading-the-summary-statistic-file">Reading the summary statistic file<a class="headerlink" href="#reading-the-summary-statistic-file" title="Permanent link">&para;</a></h1>
<p><strong>GIANT.height.gz</strong> is compressed. To read its content, you can type:</p>
<div class="codehilite"><pre><span></span>gunzip -c GIANT.height.gz <span class="p">|</span> head
</pre></div>

<p>which will shows the first 10 lines of the file</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In general, we like to work with compressed files to reduce storage space requirements</p>
</div>
<p>The <strong>GIANT.height.gz</strong> file contains the following columns:</p>
<table>
<thead>
<tr>
<th align="center">SNP</th>
<th align="center">CHR</th>
<th align="center">BP</th>
<th align="center">A1</th>
<th align="center">A2</th>
<th align="center">MAF</th>
<th align="center">SE</th>
<th align="center">P</th>
<th align="center">N</th>
<th align="center">INFO</th>
<th align="center">OR</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>With each column corresponds to the following</p>
<ul>
<li><strong>SNP</strong>: SNP ID, usually in the form of RS-ID</li>
<li><strong>CHR</strong>: The chromosome of which the SNP resides on</li>
<li><strong>BP</strong>: Chromosomal coordinate of the SNP</li>
<li><strong>A1</strong>: The effective alllele of the SNP</li>
<li><strong>A2</strong>: The non-effective allele of the SNP</li>
<li><strong>MAF</strong>: The minor allele frequency of the SNP</li>
<li><strong>SE</strong>: The standard error of the effect size esimate</li>
<li><strong>P</strong>: The P-value of association between the genotype of the SNP and the phenotype of interest</li>
<li><strong>N</strong>: Number of samples used to obtain the effect size estimate</li>
<li><strong>INFO</strong>: Usually the imputation information score. </li>
<li><strong>OR</strong>: The effect size estimate of the SNP. Can also be BETA</li>
</ul>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Some GWAS results files do not make clear which allele is the effect allele and which the non-effect allele. 
If the incorrect assumption is made in computing the PRS, then the effect of the PRS in the target data will be in the wrong direction, and so to avoid misleading conclusions it is critical that the effect allele from the base (GWAS) data is known.</p>
</div>
<h1 id="removing-duplicated-snps">Removing duplicated SNPs<a class="headerlink" href="#removing-duplicated-snps" title="Permanent link">&para;</a></h1>
<p>While it is raw, duplciated SNPs can sometimes be found in your base file.
As most PRS software do not allow duplicated SNPs in the base input, and to avoid confusion, 
it might be beneficial to remove the duplicated SNPs from the base file. </p>
<div class="codehilite"><pre><span></span>gunzip -c GIANT.height.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;{ print $1}&#39;</span> <span class="p">|</span><span class="se">\</span>
sort <span class="p">|</span><span class="se">\</span>
uniq -d &gt; duplicated.snp
</pre></div>

<p>Briefly, the above command does the following:</p>
<ol>
<li>Decompress and read the <strong>GIANT.height.gz</strong> file</li>
<li>Print out the first column of the file (which contains the SNP ID, change <code>$1</code> to other number if the SNP ID is located in another column, e.g. <code>$3</code> if the SNP ID is located on the third column)</li>
<li>Sort the SNP IDs. This will put duplicated SNP IDs next to eachother</li>
<li>Print out any duplicated SNP IDs using the uniq command and print it to the <em>duplicated.snp</em> file</li>
</ol>
<details class="note"><summary>How many duplicated SNPs are there?</summary><p>There are a total of <code>100</code> duplicated SNPs</p>
</details>
<p>Duplicated SNPs can then be removed using the <code>grep</code> command:
<div class="codehilite"><pre><span></span>gunzip -c GIANT.height.gz  <span class="p">|</span><span class="se">\</span>
grep -vf duplicated.snp <span class="p">|</span><span class="se">\</span>
gzip - &gt; Height.gz
</pre></div>
The above script does the following:
1. Decompress and read the <strong>GIANT.height.gz</strong> file 
2. Find if any row contains entries observed in <code>duplicated.snp</code> and remove them
3. Compress and write the results to <strong>Height.gz</strong></p>
<h1 id="filtering-snps-with-low-info-score-or-maf">Filtering SNPs with low INFO score or MAF<a class="headerlink" href="#filtering-snps-with-low-info-score-or-maf" title="Permanent link">&para;</a></h1>
<p>SNPs with low minor allele frequency (MAF) or imputation information score (INFO) are more likely to harbor false positives. 
It is therefore beneficial to remove SNPs with low MAF and INFO.
This is acheived by the following:</p>
<div class="codehilite"><pre><span></span>gunzip -c Height.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;NR==1 || ($6 &gt; 0.05 &amp;&amp; $6 &lt; 0.95) &amp;&amp; ($10 &gt; 0.8) {print}&#39;</span> <span class="p">|</span><span class="se">\</span>
gzip - &gt; Height.QC.gz
</pre></div>

<ol>
<li>Decompress and read the <strong>Height.gz</strong> file</li>
<li>Print the header line (<code>NR==1</code>)</li>
<li>Print any line with MAF above 0.05 and less than 0.95 (<code>$6</code> because the sixth column of the file contains the MAF information)</li>
<li>Print any line with INFO above 0.8 (<code>$10</code> because the tenth column of the file contains the INFO information)</li>
<li>Compress and write the result to <strong>Height.QC.gz</strong></li>
</ol>
<p>The <strong>Height.QC.gz</strong> file can then be used for downstream analyses</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../target/" class="btn btn-neutral float-right" title="2. Target Genotype QC">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Overview"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../target/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
