First step in Polygenic Risk Score (PRS) analyses is to obtain the GWAS summary statistics. 

In this example, we will use a modified version of the Height GWAS summary statistics generated by the [GIANT consortium](https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files#GWAS_Anthropometric_2014_Height)

# Obtaining the summary statistic file
You can download the summary statistic file [here](https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz) or use the following script:
``` bash

curl https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz -L -O

```

which will create a file call **GIANT.height.gz** in your current directory. 

A common problem is that the downloaded file can be corrupted, 
which can generate various error messages in the down-stream analyses. 
To avoid un-necessary waste of time, it is generally a good practice to 
check if the file is intact. If the `md5sum` hash is provided for the 
original file, you can check the file integrity by performing the 
`md5sum` check:

```bash tab="Linux"
md5sum GIANT.height.gz
```

```bash tab="OS X"
md5 GIANT.height.gz
```

if the file is intact, `md5sum` should generate a string of characters: `c79734b099cea663d2808bfde2e9a422`. 
If a different string is generated, the file is corrupted 

# Reading the summary statistic file
**GIANT.height.gz** is compressed. To read its content, you can type:

```bash
gunzip -c GIANT.height.gz | head
```

which will shows the first 10 lines of the file

!!! note
    Working with compressed files reduces the storage space requirements

The **GIANT.height.gz** file contains the following columns:

|SNP|CHR|BP|A1|A2|MAF|SE|P|N|INFO|OR|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|rs2073813	|1	|753541	|A	|G	|0.125	|0.0083	|0.68	|69852	|0.866425782879888	|0.996605773454898|
rs12562034	|1	|768448	|A	|G	|0.092	|0.0088	|0.55	|88015	|0.917520990188678	|0.994714020220009|
rs2980319	|1	|777122	|A	|T	|0.125	|0.006	|0.65	|148975	|0.847126999058955	|0.997303641721713|

With each column corresponds to the following

- **SNP**: SNP ID, usually in the form of RS-ID
- **CHR**: The chromosome of which the SNP resides on
- **BP**: Chromosomal coordinate of the SNP
- **A1**: The effective alllele of the SNP
- **A2**: The non-effective allele of the SNP
- **MAF**: The minor allele frequency of the SNP
- **SE**: The standard error of the effect size esimate
- **P**: The P-value of association between the genotype of the SNP and the phenotype of interest
- **N**: Number of samples used to obtain the effect size estimate
- **INFO**: Usually the imputation information score. 
- **OR**: The effect size estimate of the SNP. Can also be BETA

!!! Important
    Some GWAS results files do not make clear which allele is the effect allele and which the non-effect allele. 
    If the incorrect assumption is made in computing the PRS, then the effect of the PRS in the target data will be in the wrong direction.
    
    To avoid misleading conclusions it is critical that the effect allele from the base (GWAS) data is known.

# Removing duplicated SNPs
While it is rare, duplciated SNPs can sometimes be found in your base file.
Most PRS software do not allow duplicated SNPs in the base input, therefore it
is beneficial to remove the duplicated SNPs from the base file. 

```bash
gunzip -c GIANT.height.gz |\
awk '{ print $1}' |\
sort |\
uniq -d > duplicated.snp
```

Briefly, the above command does the following:

1. Decompress and read the **GIANT.height.gz** file
2. Print out the first column of the file (which contains the SNP ID, change `$1` to other number if the SNP ID is located in another column, e.g. `$3` if the SNP ID is located on the third column)
3. Sort the SNP IDs. This will put duplicated SNP IDs next to eachother
4. Print out any duplicated SNP IDs using the uniq command and print it to the *duplicated.snp* file


??? note "How many duplicated SNPs are there?"
    There are a total of `13` duplicated SNPs

Duplicated SNPs can then be removed using the `grep` command:
```bash
gunzip -c GIANT.height.gz  |\
grep -vf duplicated.snp |\
gzip - > Height.gz
```

The above script does the following:

1. Decompress and read the **GIANT.height.gz** file 
2. Find if any row contains entries observed in `duplicated.snp` and remove them
3. Compress and write the results to **Height.gz**

# Filtering SNPs with low INFO score or MAF 
SNPs with low minor allele frequency (MAF) or imputation information score (INFO) are more likely to harbor false positives. 
It is therefore beneficial to remove SNPs with low MAF and INFO.
This can be acheived using the following code:

```bash
gunzip -c Height.gz |\
awk 'NR==1 || ($6 > 0.05 && $6 < 0.95) && ($10 > 0.8) {print}' |\
gzip  > Height.QC.gz
```

1. Decompress and read the **Height.gz** file
2. Print the header line (`NR==1`)
3. Print any line with MAF above 0.05 and less than 0.95 (`$6` because the sixth column of the file contains the MAF information)
4. Print any line with INFO above 0.8 (`$10` because the tenth column of the file contains the INFO information)
5. Compress and write the result to **Height.QC.gz**

The **Height.QC.gz** file can then be used for downstream analyses

    
